### PR TITLE
Networking layer using rxswift and argo

### PR DESCRIPTION
**Pull request by m-revetria**

### CODE REVIEW

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 1]
```diff
@@ -9,7 +9,7 @@ author_id: remer
...
+Networking could be really a mess in big projects. If we don't take care of it we can finish with tons of networking calls made in every class of our project, accessing directly to the networking library. This may be a real problem when we try to add common behavior to every call, like logging errors, handling authentication, retries, etc. Also, we will end up by doing a lot of repetitive code as parsing services' response and checking errors. Composing a few services in order to get a final and more complex result could be a hard task. Later, understand what the code is doing won't be so easy even when reading our own code.
```
Comment (mats-claassen) [inline _posts/2016-03-07-networking-with-rxswift.md]: 'Later, understand what the code ...' -> 'Later, to understand what the code ...'

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 2]
```diff
@@ -28,12 +28,12 @@ As JSON parsing library we are going to use [Argo](https://github.com/thoughtbot
```
Comment (mats-claassen) [inline _posts/2016-03-07-networking-with-rxswift.md]: Line 25: 'external helper library that make easy integrate it with Alamofire' -> 'external helper library that makes it easy to integrate it with Alamofire'

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 3]
```diff
@@ -0,0 +1,383 @@
+---
+layout: post
+title:  "Networking Architecture with RxSwift"
+date:   2016-03-01 12:00:00
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: The date doesn't match with the file name date

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+author: Miguel Revetria
+categories: Swift,RxSwift,Alamofire,Argo,ReactiveX,Networking
+author_id: remer
+---
+
+## Motivation
+
+Networking could be really a mess in big projects. If we don't take care of it we can finish with tons of networking calls made in every class of our project, accessing directly to the networking library. This may be a real problem when we try to add common behavior to every call, like logging errors, handling authentication, retries, etc. Also, we will end up by doing a lot of repetitive code as parsing services' response and checking errors. Composing a few services in order to get a final and more complex result could be a hard task. Later, understand what the code is doing won't be so easy even when reading our own code.
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: A comma after "If we don't take care of it" would be cool

Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: "will end up doing" (without the "by")

Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: "understand**ing** what our code is doing won't be so easy**,** even when reading our own code."

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+* Define a single coupling point between our own code and the networking library used.
+* Allow us to define common behavior attached to all or some of the networking calls.
+* Allow us to compose services in an easy and understandable way.
+* Avoid repetitive code.
+
+By accomplishing previous objectives, we will be also improving our code's properties, like its readability, maintainability and modularity.
+
+## Designing our networking layer
+
+The *de facto* networking library used on Swift projects is, without any doubt, [Alamofire](https://github.com/Alamofire/Alamofire). On the other hand, the implementation of ReactiveX on Swift is [RxSwift](https://github.com/ReactiveX/RxSwift) which also has an external helper library that make easy integrate it with Alamofire. So the bases for our REST networking layer is solved.
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: "the bases **of**" sounds more fluid

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 6]
```diff
@@ -0,0 +1,383 @@
...
+* Allow us to compose services in an easy and understandable way.
+* Avoid repetitive code.
+
+By accomplishing previous objectives, we will be also improving our code's properties, like its readability, maintainability and modularity.
+
+## Designing our networking layer
+
+The *de facto* networking library used on Swift projects is, without any doubt, [Alamofire](https://github.com/Alamofire/Alamofire). On the other hand, the implementation of ReactiveX on Swift is [RxSwift](https://github.com/ReactiveX/RxSwift) which also has an external helper library that make easy integrate it with Alamofire. So the bases for our REST networking layer is solved. 
+
+As JSON parsing library we are going to use [Argo](https://github.com/thoughtbot/Argo). It has some nice features that make it a better choice than other available libraries like [ObjectMapper](https://github.com/Hearst-DD/ObjectMapper) and [Decodable](https://github.com/Anviking/Decodable). Some of its features are:
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: "As **the** JSON parsing"

Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: Maybe here it's better: "Some of **these** features are"

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 7]
```diff
@@ -0,0 +1,383 @@
...
+    
+}
+
+{% endhighlight %}
+
+> `DBLog` is a custom function that we use to include in our projects to log to the console just when debugging.
+
+### 2. Service endpoints specification
+
+Next, we are going to define our service endpoints. As Alamofire's documentation suggests we are going to make an `Enum` that implements `URLRequestConvertible` protocol but first, we are going to create some generic types that will save us some time later:
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: The "but first" use here doesn't convince me. In my opinion is better if we have it in a new sentence, or after the comma. In [Linguee](http://www.linguee.es/espanol-ingles/search?source=auto&query=but+first) there are some samples.

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 8]
```diff
@@ -0,0 +1,383 @@
...
+
+{% highlight swift %}
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: btw, I have seen that [Redcarpet Markdown can be used with our Jekyll](https://github.com/xmartlabs/xmartlabs.github.io/blame/debug-annotation-processor/_posts/2016-03-21-Debugging-an-Annotator-Processor-in-your-project.md#L8)!

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+    
+}
+
+{% endhighlight %}
+
+> We're using the public GitHub's API as an example
+
+### 3. Parsing services response
+
+Alamofire's responses are parsed easily from JSON to dictionaries, but we want to go further and use swift objects in our code. In order to parse JSON to our data models we will implement a custom `response` method that will return an `Argo.Decodable` instance.
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: I think it'd nice if you introduce other connectives than "but", like "however", "nevertheless". If it's useful to you, sometimes I use the page [English Connectives](http://www.english-on-the-web.de/vconnect.htm) when writing in English.

Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: should "swift" go capitalized here?

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+            }
+        }
+        return response(responseSerializer: responseSerializer, completionHandler: completionHandler)
+    }
+
+}
+
+{% endhighlight %}
+
+> Note that we will need an additional function in order to get an Array response from the service. Its implementation is quite similar to previous declaration.
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: "to **the** previous declaration"

Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: I don't know about Swift, but an implementation being similar to a declaration doesn't sound to me. Maybe one of these objects is not the most proper one.

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+
+### 4. RxSwift integration
+
+In order to work with Rx API we have to craete an Observable instance that will perform the networking call. We can make an extension to `Alamofire.Request` that will create an observable instance. This observable will call to our `NetworkManager`.
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: typo: "craete"

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+            .doOnError(NetworkManager.generalErrorHandler)
+    }
+
+}
+
+{% endhighlight %}
+
+### 5. Invoking our networking layer
+
+With our networking layer implemented, we are in conditions to start performing networking requests in an simple way. Take a look to the code snippet belown:
```
Comment (bryant1410) [inline _posts/2016-03-07-networking-with-rxswift.md]: typo: "belown". Btw, where editor are you writing with? Maybe a grammar checker could help you with typos

**Comment by bryant1410** [comment]
Great article! And very clear, above all.

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+
+}
+
+{% endhighlight %}
+
+## Conclusions
+
+Alamofire is the best networking library available for Swift, but this doesn't mean that a new one and better could appear in the future. Because we encapsulated all accesses to Alamofire in our `NetworkManager` and in our NetworkRouteType types, replacing Alamofire with other networking library wouldn't be a hard task.
+
+By using RxSwift we found an easy way to combine functions and compose differences services in order to get a final result. Also, its functions `doOn`, `doOnError`, `doOnNext` help us to add common behaviors to all networking calls.
```
Comment (mcousillas6) [inline _posts/2016-03-07-networking-with-rxswift.md]: "And compose differences services" -> "And compose **different** services"

File: _posts/2016-03-07-networking-with-rxswift.md [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+            return params.toDictionary()
+        }
+        return nil
+    }
+    
+}
+
+{% endhighlight %}
+
+> We're using the public GitHub's API as an example
```
Comment (mats-claassen) [inline _posts/2016-03-07-networking-with-rxswift.md]: 'public GitHub API' or 'GitHub's public API'

**Pull request closed without merging**